### ISSUE TITLE
Avoid garbage seach results on unclean shutdown

### ISSUE THREAD

**Issue by BjarniRunar** [mailpile/Mailpile#1566]
Currently, if Mailpile shuts down uncleanly, this is almost guaranteed to result in garbage search results. Sequence of events:
1. Message is discovered, assigned an MID and added to search index.
2. Search index is incrementally updated, some keyword-to-MID mappings get written out.
3. Unclean shutdown, MID and metadata is not written to disk.
4. Mailpile restarts, assigns MID to a different message.
5. Search results for MID are combined from multiple mails.

In particular, step 4 is almost guaranteed to do the wrong thing if any new mail arrives while Mailpile is shut down, because messages are processed in reverse order.

Proposed fix:

On startup, scan the global posting list for MIDs that are not present in the metadata index. Assign tombstones to all the MIDs found this way, and mark deleted.

**Comment by BjarniRunar** [comment]
Proposal on how to implement this:

1. Add method to postinglist.GlobalPostingList which finds the largest known MID
2. In search.Index.load(), use that method after loading the global posting list to sanity check the metadata
3. Create tombstone entries for all MIDs necessary to bring Metadata up to sync with the GPL.

**Comment by JackDca** [comment]
While testing PR #2023 I noticed that mailpile.idx is often not saved to the file system until Mailpile is shut down. This seems relevant to the issue of unclean shutdowns.

Examples:

Run Mailpile, with no pre-existing pile (workdir), and create an account referencing an existing IMAP server account that contains about 200 emails. Mailpile downloads the emails, indexes them, and writes kw-journal.dat (the global posting list) within a few minutes. It does not write mailpile.idx (the metadata index) until the CLI quit command is given. This is true even if Mailpile is left running for e.g. three days and receives one or two new emails from the IMAP server in that time.

Even after a quit and restart, with new messages being received after the second start of Mailpile, an updated mailpile.idx index is not written until another quit command is entered. In that case the updated mailpile.idx is identical to the first mailpile.idx but with the metadata for the new contacts and messages appended to it.

Is this is the intended behaviour?

Data integrity in the case of unclean shutdowns might be improved if mailpile.idx were written more frequently. Are there complications related to the need to append to an encrypted file?

**Comment by BjarniRunar** [comment]
Rewriting the full metadata.idx is a rare event because it is an expensive operation. Most of the time we just append changes, but appending should be relatively frequent - not just on quit. If it's ONLY happening on quit we have a bug.

**Comment by JackDca** [comment]
I will look at this more closely by monitoring my "production" Mailpile instance. The test situation I described is not realistic.

**Comment by JackDca** [comment]
I used a shell script to monitor changes (size, date/time) in mailpile.idx on my "production" Mailpile instance, with these results.    
-  Changing a tag on even one email via the UI results in an access within about 1 second to mailpile.idx that increases its size by at least a few hundred bytes. Presumably this is an append to capture the changed tag. In one case examined closely, the change consisted of an appended line for the email with the changed tag, followed by an appended line for a "missing message" which was referenced by MSG_MID from the previous line.
- When an new, incoming email is received from the IMAP server, there is no change in mailpile.idx.    
- In both the above cases, kw-journal.dat is accessed, with an increase in size.    
- When a new email is received from the IMAP server, _there is no immediate access to mailpile.idx_.    
- Over a 24 hour period, there was one access to mailpile.idx that did not appear to be associated with UI or IMAP activity, which increased its size by 5 kB. That access happened simultaneously with the first of four accesses to kw-journal.dat.

Whether this is considered to be a bug or not, it seems that if a record is appended when a tag is changed, it could just as easily be appended when an incoming email is indexed, and that this would promote data integrity.

I'll have a look at the code with this aim in mind.

**Comment by BjarniRunar** [comment]
Oh, and the ordering issue I mentioned above should be fixed by #2023 anyway. I'm currently dogfooding your changes locally, I expect to push up tomorrow probably at which point this issue gets closed! :fireworks:

**Comment by JackDca** [comment]
@BjarniRunar - Your mods to #2023 fix a vulnerability that I failed to notice, specifically, an unclean shutdown while kw-journal.dat is being appended, so thanks for that. Also for the mods to the ghost entry and the naming of the special keys.

(I do see another way to eliminate the aforementioned vulnerability without using a wholescale scan of the internal GPL. I don't like that scan, admittedly, on mostly aesthetic grounds.)

The behaviour mentioned by me above, re the delay in appending to mailpile.idx after new mail arrives or a new mailbox is scanned, is a significant nuisance even if Mailpile eventually recovers.  e.g. from IRC 2018-04-09:     

>  ... I've been using mailpile for a little while, and I just restarted it (via the mailpile-admin script) and several emails in my inbox have disappeared. Can anyone offer any guidance on how I can try to recover them?    
> ...     
> This is crazy but over the last couple hours the emails have gradually reappeared in my inbox. I regularly see "no new mail" notifications but then I will check back and another one has shown up. As now, I think all of the lost emails are back. I'm still not sure how it happened though.
> 

The recovery is superb, but I see a way to reduce the original loss of emails on unclean shutdown with minimal performance or complexity cost. So maybe this issue should remain open, perhaps renamed.

**Comment by BjarniRunar** [comment]
Glad you like the changes!

Our opinions differ on the scan... I am much happier with a wholesale scan of the GPL! This is a relatively quick operation on in-memory data, performed only once on startup. That is a much better performance profile than slowing down every single addition to the index, as we were doing before. It's also much more robust - even if there end up being bugs in how we update the max-message-id tracking key, we end up recovering gracefully anyway.

Regarding how frequently we append to the metadata index - that would belong in a separate issue, although I am not yet convinced it should be an issue at all. Note that frequently writing to the metadata file has nontrivial performance and stability implications of its own, especially as the file grows. Once it reaches a certain size (mine is nearly 300MB), rewriting the whole thing takes a significant amount of time and Mailpile isn't saving any other data to disk while that takes place: that is a vulnerable state to be in, and not at all one I want to increase the frequency of. However, we can't just append forever, because then start-up also starts to take forever... so the file has to be rewritten more often the more often we append. There's no free lunch here.

Striking a balance on this is tricky, I spent weeks testing and retesting this and tweaking it early in the apps dev cycle and I'm hesitant to change things now. They are overall working quite well as is. But I'm keeping an eye on the feedback and will course-correct if it seems warranted.

(Long term, the solution for this is to use a different storage format for the metadata index. I have some code and a design prepared for that, but I sidelined that work because the current state "good enough" and I want to get a release out.)

I'm closing this issue - thanks for your efforts here, I'm really happy to see this fixed! Great work.
